go/store/nbs/journal_index_record.go

```diff
@@ -179,62 +182,177 @@ func validateIndexRecord(buf []byte) bool {
 	return crc(buf[:off]) == readUint32(buf[off:])
 }
 
-// processIndexRecords reads a sequence of index records from |r| and passes them to the callback. While reading records
-// it makes some basic assertions that the sequence is well-formed and indexes a contiguous region for the journal file.
-func processIndexRecords(ctx context.Context, r io.ReadSeeker, sz int64, cb func(o int64, r indexRec) error) (err error) {
-	var (
-		buf  []byte
-		off  int64
-		prev uint64
-	)
-
-	rdr := bufio.NewReader(r)
-	for off < sz {
-		// peek to read next record size
-		if buf, err = rdr.Peek(uint32Size); err != nil {
-			break
-		}
-
-		l := readUint32(buf)
-		if int64(l) > sz {
-			return fmt.Errorf("invalid record size %d for index file of size %d", l, sz)
-		}
-		if len(buf) < int(l) {
-			buf = make([]byte, l)
-		}
-		if _, err = io.ReadFull(rdr, buf); err != nil {
-			break
-		}
+type lookupMeta struct {
+	batchStart int
+	batchEnd   int
+	checkSum   uint32
+	latestHash hash.Hash
+}
 
-		// we do not zero-fill the journal index and expect
-		// only complete records that will checksum
-		if !validateIndexRecord(buf) {
-			return fmt.Errorf("failed to checksum index record at %d", off)
-		}
+const indexRecTypeSize = 1
+const (
+	indexRecChunk byte = iota
+	indexRecMeta
+)
 
-		var rec indexRec
-		if rec, err = readJournalIndexRecord(buf); err != nil {
+// processIndexRecords reads batches of chunk index lookups into the journal.
+// An index batch looks like |lookup|lookup|...|meta|. The first byte of a record
+// indicates whether it is a |lookup| or |meta|.
+func processIndexRecords(ctx context.Context, rd *bufio.Reader, sz int64, cb func(lookupMeta, []lookup, uint32) error) (err error) {
+	var batchCrc uint32
+	var batch []lookup
+	for {
+		recTag, err := rd.ReadByte()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
 			return err
-		} else if rec.start != prev {
-			return fmt.Errorf("index records do not cover contiguous region (%d != %d)", rec.end, prev)
 		}
+		switch recTag {
+		case indexRecChunk:
+			l, err := readIndexLookup(rd)
+			if errors.Is(err, io.ErrUnexpectedEOF) {
+				return nil
+			} else if err != nil {
+				return err
+			}
+			batch = append(batch, l)
+			batchCrc = crc32.Update(batchCrc, crcTable, l.a[:])
 
-		if err = cb(off, rec); err != nil {
-			return err
+		case indexRecMeta:
+			m, err := readIndexMeta(rd)
+			if errors.Is(err, io.ErrUnexpectedEOF) {
+				return nil
+			} else if err != nil {
+				return err
+			}
+			if err := cb(m, batch, batchCrc); err != nil {
+				return err
+			}
+			batch = nil
+			batchCrc = 0
+		default:
+			return fmt.Errorf("expected record to start with a chunk or metadata type tag")
```

Comment:
In some ways different, in some ways...maybe not so different? Why an error here but `ErrUnexpectedEOF` is not one? If we there could be garbage beyond where we read...